drivers/led/lp50xx.c

```diff
@@ -0,0 +1,371 @@
+	int err;
+
+	if (!device_is_ready(config->bus.bus)) {
+		LOG_ERR("%s: I2C device not ready", dev->name);
+		return -ENODEV;
+	}
+
+	if (config->num_leds > config->max_leds) {
+		LOG_ERR("%s: invalid number of LEDs %d (max %d)",
+			dev->name,
+			config->num_leds,
+			config->max_leds);
+		return -EINVAL;
+	}
+
+	/*
+	 * Since the status of the LP503x controller is unknown when entering
+	 * this function, and since there is no way to reset it, then the whole
+	 * configuration must be applied.
+	 */
+
+	/* Disable bank control for all LEDs. */
+	buf[0] = LP50XX_LED_CONFIG0;
+	buf[1] = 0;
+	buf[2] = 0;
+	if (config->chip == LP5009 || config->chip == LP5012) {
+		err = i2c_write_dt(&config->bus, buf, 2);
+	} else {
+		err = i2c_write_dt(&config->bus, buf, 3);
+	}
+	if (err < 0) {
+		return err;
+	}
+
+	/* Enable LED controller. */
+	buf[0] = LP50XX_DEVICE_CONFIG0;
+	buf[1] = CONFIG0_CHIP_EN;
+	err = i2c_write_dt(&config->bus, buf, 2);
+	if (err < 0) {
+		return err;
+	}
+
+	/* Apply configuration. */
+	buf[0] = LP50XX_DEVICE_CONFIG1;
+	buf[1] = CONFIG1_PWM_DITHERING_EN | CONFIG1_AUTO_INCR_EN
+		| CONFIG1_POWER_SAVE_EN;
+	if (config->max_curr_opt) {
+		buf[1] |= CONFIG1_MAX_CURRENT_OPT;
+	}
+	if (config->log_scale_en) {
+		buf[1] |= CONFIG1_LOG_SCALE_EN;
+	}
+
+	return i2c_write_dt(&config->bus, buf, 2);
+}
+
+static const struct led_driver_api lp50xx_led_api = {
+	.on		= lp50xx_on,
+	.off		= lp50xx_off,
+	.get_info	= lp50xx_get_info,
+	.set_brightness	= lp50xx_set_brightness,
+	.set_color	= lp50xx_set_color,
+	.write_channels	= lp50xx_write_channels,
+};
+
+#define COLOR_MAPPING(led_node_id)						\
+	const uint8_t color_mapping_##led_node_id[] =				\
+		DT_PROP(led_node_id, color_mapping);
+
+#define LED_INFO(led_node_id)							\
+	{									\
+		.label		= DT_PROP(led_node_id, label),			\
+		.index		= DT_PROP(led_node_id, index),			\
+		.num_colors	=						\
+			DT_PROP_LEN(led_node_id, color_mapping),		\
+		.color_mapping	= color_mapping_##led_node_id,			\
+	},
+
+#define LP50XX_DEVICE(n, id, nchans, bbase_reg, cbase_reg)			\
+	DT_INST_FOREACH_CHILD(n, COLOR_MAPPING)					\
+										\
+	static const struct led_info lp##id##_leds_##n[] = {			\
+		DT_INST_FOREACH_CHILD(n, LED_INFO)				\
+	};									\
+										\
+	static const struct lp50xx_config lp##id##_config_##n = {		\
+		.bus			= I2C_DT_SPEC_INST_GET(n),		\
+		.chip			= LP##id,				\
+		.max_leds		= LP##id##_MAX_LEDS,			\
+		.num_leds		= ARRAY_SIZE(lp##id##_leds_##n),	\
+		.num_channels		= nchans,				\
+		.log_scale_en		= DT_INST_PROP(n, log_scale_en),	\
+		.max_curr_opt		= DT_INST_PROP(n, max_curr_opt),	\
+		.brightness_base_reg	= bbase_reg,				\
+		.color_base_reg		= cbase_reg,				\
+		.leds_info		= lp##id##_leds_##n,			\
+	};									\
+										\
+	static uint8_t lp##id##_chan_buf_##n[nchans + 1];			\
+										\
+	static struct lp50xx_data lp##id##_data_##n = {				\
+		.dev_data	= {						\
+			.min_brightness = 0U,					\
+			.max_brightness = 100U,					\
+		},								\
```

Comment:
Again, let's not waste memory to store this information. Please use static definition where you need to check minimum/maximum brightness. It will be just fine.

Also, the plan is to increase the maximum brightness to 255 for the LED API. And then all these checks will be removed.

Reply:
OK.

Good news for the increased brightness, it will be easier to set brightness using a raw `int` instead of a percent